File: android/app/src/main/java/com/bensbible/app/util/Sm2Scheduler.kt
package com.bensbible.app.util

import kotlin.math.max
import kotlin.math.roundToInt

data class Sm2Result(
    val repetitions: Int,
    val easeFactor: Double,
    val intervalDays: Int,
    val nextReviewDate: Long
)

object Sm2Scheduler {
    fun process(quality: Int, repetitions: Int, easeFactor: Double, intervalDays: Int): Sm2Result {
        var newRepetitions = repetitions
        var newEaseFactor = easeFactor
        var newInterval = intervalDays

        if (quality >= 3) {
            newRepetitions += 1
            newInterval = when (newRepetitions) {
                1 -> 1
                2 -> 6
                else -> (intervalDays * newEaseFactor).roundToInt()
            }
            val delta = 0.1 - (5 - quality) * (0.08 + (5 - quality) * 0.02)
            newEaseFactor = max(1.3, newEaseFactor + delta)
        } else {
            newRepetitions = 0
            newInterval = 1
        }

        val cal = java.util.Calendar.getInstance().apply {
            set(java.util.Calendar.HOUR_OF_DAY, 0)
            set(java.util.Calendar.MINUTE, 0)
            set(java.util.Calendar.SECOND, 0)
            set(java.util.Calendar.MILLISECOND, 0)
            add(java.util.Calendar.DAY_OF_YEAR, newInterval)
        }
        val nextDate = cal.timeInMillis
        return Sm2Result(
            repetitions = newRepetitions,
            easeFactor = newEaseFactor,
            intervalDays = newInterval,
            nextReviewDate = nextDate
        )
    }
}

File: BensBible2/Utilities/SM2Scheduler.swift
import Foundation

struct SM2Result {
    let repetitions: Int
    let easeFactor: Double
    let intervalDays: Int
    let nextReviewDate: Date
}

struct SM2Scheduler {
    static func process(quality: Int, repetitions: Int, easeFactor: Double, intervalDays: Int) -> SM2Result {
        var newRepetitions = repetitions
        var newEaseFactor = easeFactor
        var newInterval = intervalDays

        if quality >= 3 {
            newRepetitions += 1
            switch newRepetitions {
            case 1:
                newInterval = 1
            case 2:
                newInterval = 6
            default:
                newInterval = Int((Double(intervalDays) * newEaseFactor).rounded())
            }
            let delta = 0.1 - Double(5 - quality) * (0.08 + Double(5 - quality) * 0.02)
            newEaseFactor = max(1.3, newEaseFactor + delta)
        } else {
            newRepetitions = 0
            newInterval = 1
        }

        let today = Calendar.current.startOfDay(for: Date())
        let nextDate = Calendar.current.date(byAdding: .day, value: newInterval, to: today) ?? Date()
        return SM2Result(
            repetitions: newRepetitions,
            easeFactor: newEaseFactor,
            intervalDays: newInterval,
            nextReviewDate: nextDate
        )
    }
}
